### PR TITLE
feat: Add transaction to save and destroy on `Parse.Object`

### LINKED ISSUE #922: Performing batch transactions
**Is your feature request related to a problem? Please describe.**
This was already supported in the PServer, but in JS-SDK, it limits us to only perform either save or delete but not both.

**Describe the solution you'd like**
It would be better if we could constructr a batch operation like:
```js
await Parse.Object.performTransaction(
  [
    { method: "PUT", objects: [PObject] },
    { method: "DELETE, object: PObject },
    { method: "POST", object: [PObject, PObject] }
  ],
  { transaction: true, useMasterKey: true }
);
```
and performs that sequentially.

**Describe alternatives you've considered**

**Additional context**
Also, using relations when doing batch operation is not allowed because you cannot add something to relation when it is new. I don't know if there's already a solution to this but my current implementation is to use `forkJoin([DeleteObjects(), SaveObjects()])` using RxJS.

### PR DESCRIPTION
**Pull request by vahidalizad**
## Pull Request

- Report security issues [confidentially](https://github.com/parse-community/Parse-SDK-JS/security/policy).
- Any contribution is under this [license](https://github.com/parse-community/Parse-SDK-JS/blob/alpha/LICENSE).
- Link this pull request to an [issue](https://github.com/parse-community/Parse-SDK-JS/issues?q=is%3Aissue).

## Issue
This pull request addresses issues related to performing batch transactions. 
Closes: #922

## Approach
- Added a `transaction` option to the `save` and `destroy` methods in the `ParseObject` class.
- Implemented checks to ensure objects are serializable before creating a batch, as the server cannot save children and parents simultaneously in a single transactional batch.

## Tasks
- [x] Add a `transaction` option to `save` and `destroy` in the `ParseObject` class.
- [x] Ensure that objects are serializable before initiating a batch save, as transactional operations require the batch size to match the entire input.
- [x] Add relevant tests.

### CODE REVIEW

**Review by mtrezza — APPROVED** [review]

File: src/__tests__/ParseObject-test.js [hunk 1]
```diff
@@ -2024,6 +2024,142 @@ describe('ParseObject', () => {
...
+  });
+
+  it('should fail save batch with unserializable attribute and transaction option', async () => {
+    const obj1 = new ParseObject('TestObject');
+    const obj2 = new ParseObject('TestObject');
+    obj1.set('relatedObject', obj2);
+
+    try {
+      await ParseObject.saveAll([obj1, obj2], { transaction: true });
+      expect(true).toBe(false);
```
Comment (mtrezza) [inline src/__tests__/ParseObject-test.js]: Could you rewrite this and similar lines with `await expectAsync().toBeRejectedWith(...)`?

**Review by mtrezza — CHANGES_REQUESTED** [review]

File: src/__tests__/ParseObject-test.js (hunk 1 above)
Comment (vahidalizad) [reply src/__tests__/ParseObject-test.js]: I initially intended to write it as you suggested with await expectAsync().toBeRejectedWith(...). However, I noticed that we have many instances of this type of checking throughout the tests in the repo. To maintain consistency, I followed the existing pattern.

That said, we could consider updating all of them in one go in a separate PR, as there are about 25 similar instances in other files. For now, I'll go ahead and commit the changes to fix these specific cases, and perhaps we can address the rest incrementally later.

Comment (vahidalizad) [reply src/__tests__/ParseObject-test.js]: I updated these tests, can you check them out again?

**Comment by vahidalizad** [comment]
@mtrezza Could you please check and review the changes when you have a chance?

**Review by mtrezza — APPROVED** [review]
Looks good; let's see if someone from the JS team has any feedback before we merge. Btw, would you be interested to get team to get notified when someone is requesting feedback on their PR?

Is this a fix or a feature? From the issue description it looks more like a new feature?

**Comment by vahidalizad** [comment]
> Looks good; let's see if someone from the JS team has any feedback before we merge. Btw, would you be interested to get team to get notified when someone is requesting feedback on their PR?
> 
> Is this a fix or a feature? From the issue description it looks more like a new feature?

Thanks for the feedback! I agree, this is more of a feature than a fix.
Regarding the notification for the team when someone is requesting feedback on their PR, could you clarify how that would work? I'm not sure I fully understand the suggestion.

**Comment by mtrezza** [comment]
> Regarding the notification for the team when someone is requesting feedback on their PR, could you clarify how that would work? I'm not sure I fully understand the suggestion.

You'd just be added to the review team here on GitHub. When someone submits a PR and the PR is ready for review, then you'd get a notification and if you have some time you can look over it and provide feedback. No obligation though.

**Comment by vahidalizad** [comment]
> You'd just be added to the review team here on GitHub. When someone submits a PR and the PR is ready for review, then you'd get a notification and if you have some time you can look over it and provide feedback. No obligation though.

Sure, thanks for the trust! I'd be happy to help as much as I can.

**Review by mtrezza — APPROVED** [review]

**Comment by parseplatformorg** [comment]
🎉 This change has been released in version [5.3.0-alpha.4](https://github.com/parse-community/Parse-SDK-JS/releases/tag/5.3.0-alpha.4)

**Comment by jimnor0xF** [comment]
Very nice! Been waiting for this feature for quite some time so I'll give it a try later :smiley:

**Comment by parseplatformorg** [comment]
🎉 This change has been released in version [6.0.0](https://github.com/parse-community/Parse-SDK-JS/releases/tag/6.0.0)

**Pull request merged**